oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/core/expression/OAuth2Expressions.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.core.expression;
+
+import org.springframework.security.core.Authentication;
+
+/**
+ * A contract for the OAuth2 SpEL expressions entry point
+ *
+ * @author Josh Cummings
+ * @since 5.1
+ */
+public interface OAuth2Expressions {
+	Object attribute(Authentication authentication, String name);
+
+	boolean isOAuth(Authentication authentication);
```

Comment:
What are your thoughts on the usefulness of `denyOAuthClient` (perhaps named a bit confusingly) in legacy oauth?

If that has turned out to be valuable, then I think this could, too, for the same reason, e.g. endpoints that I don't care to expose over oauth.

```java
@PreAuthorize("hasAuthority('ROLE_USER') && !isOAuth()")
```

Of course, happy to leave it out until necessary. Could be a YAGNI.

Reply:
I would leave it out for now. We can always add it later if need be but my guess is that we won't need it.